### PR TITLE
Improve module wrapFunctionCalls

### PR DESCRIPTION
**Pull request by ptaeuber**
See [ticket:3812](https://trac.openmodelica.org/OpenModelica/ticket/3812)

- constant cse-variables are stored in globalKnownVars and **only
  evaluated once** before simulation instead of in each simulation step
- SimEqns for parameters are created from the globalKnownVars
- constant variables (eg. cse-variables) are solved before initialization in
  updateBoundParameters in *_08bnd.c
- primaryParameters and allPrimaryParameters are removed
- improved dumps and documentation

---

- these changes do not influence the scaling
- copying of hashtables is removed
- some other improvements and fixes

The module should not fail for models in the coverage test any more. However, there could be an effect on the coverage, so we should keep an eye on it.

### CODE REVIEW

File: Compiler/BackEnd/BackendVariable.mo [hunk 1]
```diff
@@ -1118,6 +1134,18 @@ algorithm
...
 end isStringParam;
 
+public function isVar
+"Return true if variable is a variable."
+  input BackendDAE.Var inVar;
+  output Boolean outBoolean;
+algorithm
+  outBoolean:= match (inVar)
+    case BackendDAE.VAR(varKind = BackendDAE.VARIABLE()) then true;
+    case BackendDAE.VAR(varKind = BackendDAE.DISCRETE()) then true;
```
Comment (lochel) [inline Compiler/BackEnd/BackendVariable.mo]: What about, e.g., `DUMMY_STATE`?

File: Compiler/BackEnd/BackendVariable.mo [hunk 2]
```diff
@@ -2358,6 +2404,30 @@ algorithm
...
 
+public function deleteVarIfExistsAndReturn
+"author: PA
+  Deletes a variable from Variables and returns it."
+  input DAE.ComponentRef inComponentRef;
+  input BackendDAE.Variables inVariables;
+  output list<BackendDAE.Var> outVarLst;
+  output BackendDAE.Variables outVariables;
+algorithm
+  (outVarLst,outVariables) := matchcontinue(inComponentRef,inVariables)
```
Comment (lochel) [inline Compiler/BackEnd/BackendVariable.mo]: It's better to use `try` instead of dummy match patterns.

**Comment by OpenModelica-Hudson** [comment]
The tests run correctly according to [OpenModelica_TEST_PULL_REQUEST 2017-04-11_14-44-42](https://test.openmodelica.org/hudson//job/OpenModelica_TEST_PULL_REQUEST/4188/console).

**Comment by casella** [comment]
Just a naive question: are you making sure that impure function calls are not turned into constant cse variables?

**Comment by lochel** [comment]
Yes. Impure function calls are not touched by wrapFunctionCalls.

**Pull request closed without merging**